rust/src/ssh/parser.rs

```diff
@@ -0,0 +1,187 @@
+/* Copyright (C) 2018 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+//nom5 use nom::bytes::complete::is_not;
+use nom::{be_u32, be_u8, rest};
+
+#[inline]
+fn is_not_lineend(b: u8) -> bool {
+    if b == 10 || b == 13 {
+        return false;
+    }
+    return true;
+}
+
+//may leave \r at the end to be removed
+named!(pub ssh_parse_line<&[u8], &[u8]>,
+    terminated!(
+        take_while!(is_not_lineend),
+        alt!( tag!("\n") | tag!("\r\n") |
+              do_parse!(
+                    bytes: tag!("\r") >>
```

Comment:
I am not sure that SSH parser should accept single CR as end of banner, but that is what is being done by the C parser

Reply:
I suppose this is something to test against real implementations.

Reply:
I have been looking at openSSH only, which accepts LF and CRLF as end of line/banner (but not CR only) cf https://github.com/openssh/openssh-portable/blob/master/ssh_api.c#L345

The current unit tests make sure CR only must be considered as end of line/banner

Which other real implementations should I look at ?
Should this be part of this PR ? Or should I create a new ticket for it ?

Reply:
Ticket sounds fine.

Reply:
https://redmine.openinfosecfoundation.org/issues/3485